crates/lsp/src/documents.rs

```diff
@@ -119,78 +125,69 @@ impl Document {
+                self.line_index = LineIndex::new(&self.contents);
+            }
+            last_start_line = range.start.line;
 
-        self.parse = parse;
-        self.contents = contents;
-        self.line_index.index =
-            triomphe::Arc::new(biome_line_index::LineIndex::new(&self.contents));
-        self.version = Some(new_version);
-    }
+            // This is a panic if we can't convert. It means we can't keep the document up
+            // to date and something is very wrong.
+            let range: Range<usize> =
+                from_proto::text_range(range, &self.line_index.index, self.position_encoding)
+                    .expect("Can convert `range` from `Position` to `TextRange`.")
+                    .into();
 
-    // --- source
-    // authors = ["rust-analyzer team"]
-    // license = "MIT OR Apache-2.0"
-    // origin = "https://github.com/rust-lang/rust-analyzer/blob/master/crates/rust-analyzer/src/lsp/utils.rs"
-    // ---
-    fn apply_document_changes(
-        encoding: PositionEncoding,
-        file_contents: &str,
-        mut content_changes: Vec<lsp_types::TextDocumentContentChangeEvent>,
-    ) -> String {
-        // If at least one of the changes is a full document change, use the last
-        // of them as the starting point and ignore all previous changes.
-        let (mut text, content_changes) = match content_changes
-            .iter()
-            .rposition(|change| change.range.is_none())
-        {
-            Some(idx) => {
-                let text = std::mem::take(&mut content_changes[idx].text);
-                (text, &content_changes[idx + 1..])
-            }
-            None => (file_contents.to_owned(), &content_changes[..]),
-        };
-        if content_changes.is_empty() {
-            return text;
+            self.contents
+                .replace_range(range, &line_ending::normalize(change.text));
         }
 
-        let mut line_index = biome_line_index::LineIndex::new(&text);
-
-        // The changes we got must be applied sequentially, but can cross lines so we
-        // have to keep our line index updated.
-        // Some clients (e.g. Code) sort the ranges in reverse. As an optimization, we
-        // remember the last valid line in the index and only rebuild it if needed.
-        // The VFS will normalize the end of lines to `\n`.
-        let mut index_valid = !0u32;
-        for change in content_changes {
-            // The None case can't happen as we have handled it above already
-            if let Some(range) = change.range {
-                if index_valid <= range.end.line {
-                    line_index = biome_line_index::LineIndex::new(&text);
-                }
-                index_valid = range.start.line;
-                if let Ok(range) = from_proto::text_range(range, &line_index, encoding) {
-                    text.replace_range(Range::<usize>::from(range), &change.text);
-                }
```

Comment:
We used to just...skip (??) the file update if we were unable to convert from `Position` to `TextRange`.

This is a critical failure IMO and should be treated as a panic, which we now do